lib/ecto/changeset.ex

```diff
@@ -506,10 +583,17 @@ defmodule Ecto.Changeset do
   """
   @spec put_change(t, atom, term) :: t
   def put_change(%Changeset{} = changeset, key, value) do
-    update_in changeset.changes, &put_change(changeset.model, &1, key, value)
+    changeset = ensure_has_types(changeset)
+    type = Map.get(changeset.types, key)
+    update_in changeset.changes, &put_change(changeset.model, &1, key, value, type)
+  end
+
+  defp put_change(_model, acc, key, value, {:embed, embed}) do
+    value = Ecto.Embedded.wrap_change(embed, value)
+    Map.put(acc, key, value)
```

Comment:
Can we do something smarter here?
For embed_one it would be easy to check for changes, but for embed_many it will be hard...